### ISSUE TITLE
[REQ] Alternative installation methods for the android app

### ISSUE THREAD

**Issue by Axolord** [timschneeb/GalaxyBudsClient#582]
I am using a degoogled android phone (with CalyxOS) and would like to use your application to update my buds firmware.

Using apps from the play store is possible via the Aurora Store, but installing paid apps only works in conjunction with a Google Accound (that pays for the app) and I do not want to do that.

Would it be possible and acceptable for you to provide the APK either via Github releases (so that obtainium can install it) or maybe even F-Droid? You could ask users to donate, instead of paying (just like OsmAnd does it for example) or implement your own license verification system (like fairEmail).

Thanks for the app though, looks nice!

**Comment by salihefee** [comment]
You can clone the source and just build it locally. The icon of the locally built app says "Demo" but I don't know whats different on this version than the Google Play version since I haven't seen any locked or missing features.

**Comment by scynrays** [comment]
Hey, could you tell me how i could do that, i can't seem to find anything online

**Comment by salihefee** [comment]
Install .NET SDK 8, Android SDK, and the android workload for .NET and run this command in the GalaxyBudsClient.Android project directory.
`dotnet publish -f net8.0-android -c Debug -p:AndroidPackageFormats=apk -p:AndroidSdkDirectory={YOUR ANDROID SDK PATH} -p:DebugType=PdbOnly -p:EmbedAssembliesIntoApk=true`
You can also edit the .csproj to remove the demo icon and other demo specific stuff.

**Comment by salihefee** [comment]
@timschneeb Should I be telling people how to do this? I figured it would be okay since the source code is just publicly available. Also, is there anything different on the demo version?

**Comment by timschneeb** [comment]
> Should I be telling people how to do this?

Yes, that's fine. My thought process behind this was that people who compile it themselves can use it for free and people who want the convenience of prebuilt APKs and auto-updates need to purchase it.

The demo version just disables some advanced features, like firmware updates. Otherwise, you can use it normally. It's more of a 'lite' instead of a 'demo' version.

Distributing the demo version to F-Droid (most people don't need those advanced features anyway) is on my TODO list.

**Comment by salihefee** [comment]
> > Should I be telling people how to do this?
> 
> Yes, that's fine. My thought process behind this was that people who compile it themselves can use it for free and people who want the convenience of prebuilt APKs and auto-updates need to purchase it.
> 
> The demo version just disables some advanced features, like firmware updates. Otherwise, you can use it normally. It's more of a 'lite' instead of a 'demo' version.
> 
> Distributing the demo version to F-Droid (most people don't need those advanced features anyway) is on my TODO list.

I am pretty sure I can still do firmware updates with the version I compiled myself, I can access the tab and do the flashing process. (Though it does not work for me, both on the linux and android versions. I created an issue for it.) Is this how it is supposed to be?

**Comment by timschneeb** [comment]
> Is this how it is supposed to be?

Yes, you can also pass the NotDemo=true build parameter to dotnet publish instead of editing the csproj file.

**Comment by salihefee** [comment]
> > Is this how it is supposed to be?
> 
> Yes, you can also pass the NotDemo=true build parameter to dotnet publish instead of editing the csproj file.

Oh, I did not know how that parameter was passed to the .csproj so I edited it myself lol

**Comment by Meow-pilli** [comment]
> Install .NET SDK 8, Android SDK, and the android workload for .NET and run this command in the GalaxyBudsClient.Android project directory. `dotnet publish -f net8.0-android -c Debug -p:AndroidPackageFormats=apk -p:AndroidSdkDirectory={YOUR ANDROID SDK PATH} -p:DebugType=PdbOnly -p:EmbedAssembliesIntoApk=true` You can also edit the .csproj to remove the demo icon and other demo specific stuff.

I've Tried this and it doesn't work, is there any other way to get demo app on F-droid?

**Comment by salihefee** [comment]
What exactly do you mean by "it doesn't work"? What error do you get?

**Comment by NovigradWitcher** [comment]
> Install .NET SDK 8, Android SDK, and the android workload for .NET and run this command in the GalaxyBudsClient.Android project directory.
> `dotnet publish -f net8.0-android -c Debug -p:AndroidPackageFormats=apk -p:AndroidSdkDirectory={YOUR ANDROID SDK PATH} -p:DebugType=PdbOnly -p:EmbedAssembliesIntoApk=true`
> You can also edit the .csproj to remove the demo icon and other demo specific stuff.



You also need to pass the IsAndroid=true parameter (at least for me, on osx). Also Android 14 platform needs to be installed through sdkManager.

**Comment by pzmarzly** [comment]
I used Docker to compile the app without cluttering my system with Android dev dependencies. Here's the vibecoded Dockerfile, hopefully will be useful to someone else:

```dockerfile
FROM mcr.microsoft.com/dotnet/sdk:10.0 AS build

# Install JDK 17 (required by Android SDK tools) and utilities
RUN apt-get update \
    && apt-get install -y --no-install-recommends openjdk-17-jdk-headless wget unzip \
    && rm -rf /var/lib/apt/lists/*

ENV JAVA_HOME=/usr/lib/jvm/java-17-openjdk-amd64
ENV ANDROID_HOME=/opt/android-sdk
ENV PATH="${ANDROID_HOME}/cmdline-tools/latest/bin:${ANDROID_HOME}/platform-tools:${PATH}"

# Install .NET SDK 10.0.103 (required by global.json rollForward: latestPatch)
RUN wget -q https://dot.net/v1/dotnet-install.sh -O /tmp/dotnet-install.sh \
    && chmod +x /tmp/dotnet-install.sh \
    && /tmp/dotnet-install.sh --version 10.0.103 --install-dir /usr/share/dotnet \
    && rm /tmp/dotnet-install.sh

# Install Android SDK command-line tools
RUN mkdir -p ${ANDROID_HOME}/cmdline-tools \
    && wget -q https://dl.google.com/android/repository/commandlinetools-linux-11076708_latest.zip -O /tmp/cmdline-tools.zip \
    && unzip -q /tmp/cmdline-tools.zip -d /tmp/cmdline-tools-unzipped \
    && mv /tmp/cmdline-tools-unzipped/cmdline-tools ${ANDROID_HOME}/cmdline-tools/latest \
    && rm -rf /tmp/cmdline-tools.zip /tmp/cmdline-tools-unzipped

# Accept Android SDK licenses and install required platform components
RUN yes | sdkmanager --licenses > /dev/null 2>&1 \
    && sdkmanager "platforms;android-36" "build-tools;36.0.0" "platform-tools"

# Install .NET Android workload
RUN dotnet workload install android

# Copy source
WORKDIR /src
COPY . .

# Build the APK
RUN dotnet publish -f net10.0-android36.1 -c Debug \
    -p:AndroidPackageFormats=apk \
    -p:AndroidSdkDirectory=${ANDROID_HOME} \
    -p:DebugType=PdbOnly \
    -p:EmbedAssembliesIntoApk=true \
    -p:IsAndroid=true \
    -p:NoDemo=false \
    GalaxyBudsClient.Android/GalaxyBudsClient.Android.csproj

# Output stage — just the APK
FROM scratch AS output
COPY --from=build /src/GalaxyBudsClient.Android/bin/Debug/net10.0-android36.1/publish/me.timschneeberger.galaxybudsclient.demo-Signed.apk /
```

By the time you are reading it, you will probably need to update versions of all tools in that Dockerfile.

Usage:

```bash
docker build -f Dockerfile.android --output type=local,dest=/tmp .
adb install ./tmp/me.timschneeberger.galaxybudsclient.demo-Signed.apk
```

And most importantly, go send some money to @timschneeb via https://ko-fi.com/thepbone :) I just donated €30.

If F-Droid release does not happen, it'd be really cool to have an automated .apk release in GitHub Releases - then we, the degoogled people, could at least point [Obtainium](https://github.com/ImranR98/Obtainium) at it.

**Comment by tempneq2000** [comment]
I second that. I hope there is a non-Google installation method. Payment can be settled in another way similar to FairEmail or FUTO. Doesn't matter even if I have to pay more. Also Google is eating up 30% anyways
